qforce/forcefield.py

```diff
@@ -387,6 +391,236 @@ def set_charge(self, non_bonded):
+                frcmod.write(f"{unique_at[ids[0]][atom_ids[ids[0]]]:<2}-")  #Atom 1
+                frcmod.write(f"{unique_at[ids[1]][atom_ids[ids[1]]]:<2}-")  #Atom 2
+                frcmod.write(f"{unique_at[ids[2]][atom_ids[ids[2]]]:<2}-")  #Atom 3
+                frcmod.write(f"{unique_at[ids[3]][atom_ids[ids[3]]]:<2}")   #Atom 4
+                frcmod.write(f" 1 {c[0]:>15.2f} {0.00:>15.2f}  1\n")            #IDIVF=1, fcte, angle, n=1 (n*angle)
+
+
+            # improper dihedrals
+            if len(terms['dihedral/improper']) > 0:
+                frcmod.write("\nIMPROPER\n")
+                for dihed in terms['dihedral/improper']:
+                    ids = dihed.atomids + 1
+                    equ = np.degrees(dihed.equ)           # Degrees -> Degrees
+                    fconst = dihed.fconst * (0.239005)/2  # kJ/mol -> kcal/mol -> f/2 as is Vn/2 in amber
+                    frcmod.write(f"{unique_at[ids[0]][atom_ids[ids[0]]]:<2}-")  #Atom 1
+                    frcmod.write(f"{unique_at[ids[1]][atom_ids[ids[1]]]:<2}-")  #Atom 2
+                    frcmod.write(f"{unique_at[ids[2]][atom_ids[ids[2]]]:<2}-")  #Atom 3
+                    frcmod.write(f"{unique_at[ids[3]][atom_ids[ids[3]]]:<2}")   #Atom 4
+                    frcmod.write(f" 1 {fconst:>15.2f} {equ:>15.2f} 2\n")        #IDIVF=1, fcte, angle, n=2 (n*angle)
+
+    def write_frcmod_nonbond(self, frcmod, non_bonded, atom_ids, unique_at):
+        gro_atomtypes, gro_nonbonded, gro_1_4 = self.convert_to_gromacs_nonbonded(non_bonded)
+        frcmod.write("\nNONB\n")
+        # support to gaff and gaff2
+        if self.comb_rule == 2:
+            for i in range(1,self.n_atoms+1,1):
+                frcmod.write(f"{unique_at[i][atom_ids[i]]:<8}")
+                frcmod.write(f"{gro_atomtypes[atom_ids[i]][0]*5.612:>12.4f}")
+                frcmod.write(f"{gro_atomtypes[atom_ids[i]][1]/4.184:>12.4f}")
+                frcmod.write(f"\t{atom_ids[i]}\n")
+
+
+##############################################
+#           GET ATOM TYPES AND IDS           #
+##############################################
+    def get_atom_types(self, topo, non_bonded):
+        atom_ids={}  #dictonary containing original atom types
+        unique_at={} #dictonary containing new unique atom types
+        unique_masses={}
+
+        ascii_lowercase = list(string.ascii_lowercase)
+        ascii_digits = list(string.digits)
+        ascii_uppercase = list(string.ascii_uppercase)
+
+        for i, (lj_type, a_name) in enumerate(zip(non_bonded.lj_types, self.atom_names), start=1):
+            atom_ids[i]=lj_type
+            if self.n_atoms < 36:
+                if i <= 10:
+                    unique_at[i]={atom_ids[i] : "Q{}".format(ascii_digits[i-1])}
+                elif i > 10:
+                    unique_at[i]={atom_ids[i] : "Q{}".format(ascii_lowercase[i-11])}
+            else:
+                unique_at[i]={atom_ids[i] : "{}{}".format(ascii_uppercase[i-1], ascii_lowercase[i-1])}
```

Comment:
could you explain what is happening here? I didn't fully understand what you did with the ascii characters and why you needed that

Reply:
For Amber, if there are two CA-CA bonds, the second CA-CA entry overwrites the parameters for the first.
Because of that, we used ascii characters to create unique atom types for each bond